### PR TITLE
[PlutusTx] [Test] 'Eq' for 'Value'

### PR DESCRIPTION
**Pull request by effectfully**
This adds some budget tests for `(==) @Value`.

Needs docs, don't look here yet.

### CODE REVIEW

**Review by effectfully — COMMENTED** [review]
After redoing it several times the current state of the PR is what I settled on. 

No property tests, because so far I'm only concerned with making sure that whatever I implement as a better version of `(==) @Value` is indeed faster. Plus I need to benchmark my version against not only the existing one, but also the winner from #5135 (`valueEqualsValue4`). I hope mine's gonna be faster, but you never know.

**Comment by effectfully** [comment]
This is ready for review, but

1. CI throws random nix errors that I don't understand
2. there are some true errors related to name duplication etc, but it's hard to dig them out because of 1
3. I need to create golden files per GHC version, which I just realized

I'll handle 2 and 3 and they shouldn't really affect reviews, so please review. I'll need help with 1, I've no idea what's going on there.

File: plutus-tx/src/PlutusTx/AssocMap.hs [hunk 1]
```diff
@@ -149,6 +152,10 @@ instance (Pretty k, Pretty v) => Pretty (Map k v) where
 fromList :: [(k, v)] -> Map k v
 fromList = Map
 
+{-# INLINEABLE fromListSafe #-}
+fromListSafe :: Eq k => [(k, v)] -> Map k v
```
Comment (michaelpj) [inline plutus-tx/src/PlutusTx/AssocMap.hs:156]: unused?

**Comment by michaelpj** [comment]
It's a bit hidden, but you can eventually find that it's a failure to solve for the dependencies on 8.10: https://ci.iog.io/build/1252601/nixlog/1

This is because the plugin is not buildable on < 9.2, so we now can't build the `plutus-ledger-api` tests on 8.10, which we were previously able to do. I'm unsure about that - thus far we've kept `plutus-ledger-api` entirely free of the plugin. But annoyingly `Value` is defined there, so not sure what else to do. We could either just accept that the tests don't work on 8.10, or make another test component that has stuff that depends on the plugin and doesn't build on 8.10.

File: plutus-ledger-api/test/Spec/Value.hs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+haskellEqValue value1 value2 = toMap value1 Haskell.== toMap value2 where
+    toMap
+        = Map.filter (Haskell.not . Map.null)
+        . Haskell.fmap (Map.filter (Haskell./= 0))
+        . Map.fromListWith (Map.unionWith (Haskell.+))
+        . Haskell.map (Haskell.fmap $ Map.fromListWith (Haskell.+))
+        . valueToLists
+
+-- | Check whether all currencies and tokens within each of the currencies occur uniquely.
+allDistinct :: Value -> Bool
```
Comment (michaelpj) [inline plutus-ledger-api/test/Spec/Value.hs:182]: Can't you use your `safeFromList` to avoid needing this?

**Review by michaelpj — APPROVED** [review]

**Comment by michaelpj** [comment]
To be clear, to get stuff to be ignored for 8.10 you just have to do what we do elsewhere and add an GHC version conditional guarding `buildable: False`.

File: plutus-ledger-api/test/Spec/Value.hs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    (errOrRes, cost)
+        = PLC.runCekNoEmit PLC.defaultCekParameters PLC.counting
+        . PLC.runQuote
+        . PLC.unDeBruijnTermWith (Haskell.error "Free variable")
+        . PLC._progTerm
+        $ getPlc prog
+    res = either Haskell.throw id $ errOrRes >>= PLC.readKnownSelf
+
+-- | Check equality of two compiled 'Value's directly in Haskell.
+haskellEqValue :: Value -> Value -> Bool
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-ledger-api/test/Spec/Value.hs:172]: > Or is the whole point to have another implementation?

Exactly, I'm testing the actual implementation against a much simpler one, not just computing the budgets.

File: plutus-ledger-api/test/Spec/Value.hs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        Haskell.and . Map.fromListWith (\_ _ -> False) . Haskell.map (Haskell.fmap $ \_ -> True))
+    . valueToLists
+
+-- | Return all the pairs of elements of the given list.
+--
+-- > (x, y) `elem` pairs xs ==> fromJust (x `elemIndex` xs) <= fromJust (y `elemIndex` xs)
+--
+-- >>> pairs "abc"
+-- [('a','a'),('a','b'),('b','b'),('b','c'),('c','c')]
+pairs :: [a] -> [(a, a)]
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-ledger-api/test/Spec/Value.hs:196]: Nope, that would produce

```haskell
[('a','a'),('a','b'),('a','c'),('b','a'),('b','b'),('b','c'),('c','a'),('c','b'),('c','c')]
```

not

```haskell
[('a','a'),('a','b'),('b','b'),('b','c'),('c','c')]
```

I'll rename it to `pairsUnord` or something.

File: plutus-ledger-api/test/Spec/Value.hs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                    ]
+                else Pretty.group $ Pretty.pretty budget
+
+test_EqValue :: TestTree
+test_EqValue =
+    runTestNestedIn ["test", "Spec"] $
+        testNestedGhc "golden"
+            [ test_EqCurrencyList "Short" currencyListOptions
+            , test_EqCurrencyList "Long" currencyLongListOptions
+            ]
```
Comment (effectfully) [inline plutus-ledger-api/test/Spec/Value.hs:231]: I don't like these tests... They cover some interesting corner cases, but not nearly enough. I'll make a ticket for improving them.

File: plutus-tx/src/PlutusTx/AssocMap.hs (hunk 1 above)
Comment (effectfully) [reply plutus-tx/src/PlutusTx/AssocMap.hs:156]: Yep, but I believe it's worth having it if just to emphasize that the default version isn't safe.

File: plutus-ledger-api/test/Spec/Value.hs (hunk 2 above)
Comment (effectfully) [reply plutus-ledger-api/test/Spec/Value.hs]: Maybe, but the point is that it's for tests and I want an alternative implementation for testing purposes that does not rely on anything from `PlutusTx.AssocMap` and is self-contained etc.

**Comment by effectfully** [comment]
@michaelpj I've separated the new tests into a separate stanza, but CI still seems to complain. Any suggestions what I should do next?

**Comment by michaelpj** [comment]
I pushed a couple of commits:
1. Annoying Nix stuff since we need to turn off anything using the plugin when cross-compiling.
2. Some unnecessary dependencies that were causing a `-Werror` failure
